### ISSUE TITLE
Implement htsget id resolver interface

### ISSUE THREAD

**Issue by brainstorm** [chris-zen/htsget-mvp#10]
When a query is sent to htsget via, for instance, `GET /reads/<id>`, the spec says that `id` can be any of those things:

![Screen Shot 2021-04-09 at 9 37 02 pm](https://user-images.githubusercontent.com/175587/114174272-bc0e1a00-997b-11eb-8182-c3a815e2e7a5.png)

The spec is intentionally loose in this regard since different organizations might have differing requirements regarding how this endpoints and internal (organizational) structure they have in place.

A "resolver" interface should allow for those organizations to easily "plug and play" htsget into their systems more easily and/or filter what's a legal ID or not, [perhaps via regexps like the reference implementation does](https://github.com/ga4gh/htsget-refserver/blob/bd3fa58b20377c5151cfa33e45d48ceeb49f28bb/internal/htsconfig/datasources.go).

**Comment by brainstorm** [comment]
Not necessarily DRS reference implementation, but could be useful to examine: https://github.com/ncbi/ncbi-drs.

@victorskl @AndrewPattison Keep this issue in the back of your mind(s), we might need to discuss to which extent we implement this and which assumptions we do on the htsget-rs side?

**Comment by brainstorm** [comment]
For now this will entail implementing support for `regexp`-based mapping of resources:

https://github.com/ga4gh/htsget-refserver/blob/main/internal/htsconfig/datasources.go

BUT, the resolver interface (perhaps another crate named `htsget-resolver`) should be modular and accomodate other disambiguation mechanisms when trying to access a dataset. On top of the regexp mechanism outlined above, this interface could be based on URL schemes:

```
s3://...  (AWS)
gs://...  (Google)
gds://...(Illumina's ICA)
drs://... (GA4GH)
file://... (Local)
https://... (On-prem server or AWS presigned URLs)
```

**Comment by brainstorm** [comment]
Assigning this one to you, @CastilloDel, I think that @mmalenic will be busy with the S3 storage backend modeling since [I've introduced a couple of extra requirements there](https://github.com/umccr/htsget-rs/issues/9#issuecomment-897629442).

**Comment by CastilloDel** [comment]
I'm thinking the easiest  approach would be to configure it in a way similar to how the reference implementation does it, but using environment variables. Then the htsget-http-core could provide helpers to build an id resolver struct (from a htsget-id-resolver crate) and the LocalStorage would store the id resolver and use it when it is needed.

**Comment by CastilloDel** [comment]
https://crates.io/crates/regex seems like the way to go

**Comment by brainstorm** [comment]
Yes! Sounds good, other ID resolvers are getting stabilised still in specs,
so this is a good start indeed, thanks Daniel!
